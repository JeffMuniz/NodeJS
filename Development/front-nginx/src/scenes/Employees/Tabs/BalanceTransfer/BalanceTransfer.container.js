import React, { Component } from "react";

import { connect } from "react-redux";
import { inputToMoney } from "@utils";
import { shape, string, func } from "prop-types";
import {
  setTransferConfig,
  getTransferConfig,
  getValuesTransfer,
} from "src/api/card/cardTransfer";

import { showToast as showToastAction } from "src/common/Toast/redux/actions/toast";
import FormWrapper from "src/modules/Form/Form";
import BalanceTransfer from "./BalanceTransfer";

export class BalanceTransferContainer extends Component {
  state = {
    hasBalanceTransfer: false,
    hasMealToFood: false,
    hasFoodToMeal: false,
    hasAcceptTerms: false,
    foodValue: null,
    mealValue: null,
    visibleTooltip: false,
    loading: false,
    disabledBalanceTransfer: false,
    tooltipBalanceTransfer: false,
  };

  async componentDidMount() {
    const {
      selectedCompany: { cnpj },
      user,
    } = this.props;

    const config = await getTransferConfig(cnpj, +user.id);
    const disabledBalance = this.configDisabledBalanceTransfer(config);
    this.getBalanceTransfer(cnpj, disabledBalance);
  }

  onChangeState = (label, value) => {
    if (label === "hasBalanceTransfer") {
      this.setValueHasBalanceTransfer(value);
      return;
    }

    if (label === "hasAcceptTerms") {
      this.setState({
        ...this.state,
        hasAcceptTerms: value,
      });

      return;
    }

    this.removeValue(label, value);
  };

  setValueHasBalanceTransfer = value => {
    const { disabledBalanceTransfer, tooltipBalanceTransfer } = this.state;

    if (!disabledBalanceTransfer) {
      this.setState({
        hasFoodToMeal: false,
        hasMealToFood: false,
        foodValue: "",
        mealValue: "",
        hasBalanceTransfer: value,
      });

      return;
    }

    this.setState({
      ...this.state,
      tooltipBalanceTransfer: !tooltipBalanceTransfer,
    });
  };

  getBalanceTransfer = async (cnpj, disabledBalance) => {
    const result = await getValuesTransfer(cnpj);

    if (disabledBalance) {
      this.setState({
        ...this.state,
        hasBalanceTransfer: false,
      });

      return;
    }

    this.setState({
      ...result,
      visibleTooltip: false,
      hasAcceptTerms: false,
      tooltipBalanceTransfer: disabledBalance,
    });
  };

  configDisabledBalanceTransfer = config => {
    const disabledBalance = !config.valeRefeicao || !config.valeAlimentacao;

    this.setState({
      ...this.state,
      disabledBalanceTransfer: disabledBalance,
    });

    return disabledBalance;
  };

  removeValue = (label, value) => {
    const { foodValue, mealValue } = this.state;

    const obj = {
      ...this.state,
      [label]: value,
      foodValue:
        label === "hasFoodToMeal" ? this.formatValue(value) : foodValue,
      mealValue:
        label === "hasMealToFood" ? this.formatValue(value) : mealValue,
    };

    this.setState({
      ...obj,
      hasBalanceTransfer:
        !obj.hasFoodToMeal && !obj.hasMealToFood
          ? false
          : obj.hasBalanceTransfer,
    });
  };

  formatValue = value => (value ? "0,00" : "");

  showTooltip = () => {
    const { visibleTooltip } = this.state;

    this.setState({
      ...this.state,
      visibleTooltip: !visibleTooltip,
    });
  };

  changeValue = (label, value) => {
    const newState = this.state;
    newState[label] = inputToMoney(value || 0).replace("R$", "");

    this.setState({ newState });
  };

  submit = async () => {
    this.setState({ loading: true });
    const { selectedCompany, user, showToast } = this.props;

    try {
      const balanceTransferObj = {
        ...this.state,
        cnpj: selectedCompany.cnpj,
        cpf: user.cpf,
      };

      await setTransferConfig(balanceTransferObj);

      this.setState({ loading: false });
      showToast({
        id: "success-balance-transfer-id",
        label: "Configuração realizada com sucesso!",
      });
    } catch (error) {
      this.setState({ loading: false });
      showToast({
        id: "error-balance-transfer-id",
        label: "Ocorreu um erro ao alterar Trasferência de saldo!",
      });
    }
  };

  render() {
    const {
      hasBalanceTransfer,
      hasFoodToMeal,
      hasMealToFood,
      hasAcceptTerms,
      tooltipBalanceTransfer,
      foodValue,
      mealValue,
      visibleTooltip,
      loading,
    } = this.state;

    return (
      <FormWrapper enableReinitialize>
        <BalanceTransfer
          hasBalanceTransfer={hasBalanceTransfer}
          hasFoodToMeal={hasFoodToMeal}
          hasMealToFood={hasMealToFood}
          hasAcceptTerms={hasAcceptTerms}
          tooltipBalanceTransfer={tooltipBalanceTransfer}
          foodValue={foodValue}
          mealValue={mealValue}
          visibleTooltip={visibleTooltip}
          loading={loading}
          showTooltip={() => this.showTooltip()}
          changeValue={(label, value) => this.changeValue(label, value)}
          onChangeTransfer={(label, value) => this.onChangeState(label, value)}
          onSubmit={() => this.submit()}
        />
      </FormWrapper>
    );
  }
}

export const mapStateToProps = ({ selectedCompanyTree, user }) => ({
  selectedCompany: selectedCompanyTree.selectedCompany,
  user: user.profile.data,
});

export const mapDispatchToProps = {
  showToast: showToastAction,
};

BalanceTransferContainer.propTypes = {
  selectedCompany: shape({
    cnpj: string,
  }),
  user: shape({
    cpf: string,
  }),
  showToast: func.isRequired,
};

BalanceTransferContainer.defaultProps = {
  selectedCompany: { cnpj: "" },
  user: { cpf: "" },
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(BalanceTransferContainer);
