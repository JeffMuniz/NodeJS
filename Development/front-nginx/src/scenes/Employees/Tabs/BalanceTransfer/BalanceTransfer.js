import React, { Fragment } from "react";

import { func, bool, string } from "prop-types";
import { CheckBox, Switch, Button } from "@common";
import { buttonTypes as buttonTypesEnum } from "@enums";
import { CustomField } from "src/modules/Form";

import TooltipBalanceTransfer from "./TooltipBalanceTransfer/TooltipBalanceTransfer";

import {
  Container,
  Main,
  Title,
  SwitchContainer,
  Text,
  ButtonsWrapper,
  InputContainer,
  InputWrapper,
  ButtonBalanceTransfer,
  TextWrapper,
  Relative,
} from "./BalanceTransfer.styles";

export const BalanceTransfer = ({
  hasBalanceTransfer,
  hasFoodToMeal,
  hasMealToFood,
  onChangeTransfer,
  hasAcceptTerms,
  tooltipBalanceTransfer,
  mealValue,
  foodValue,
  showTooltip,
  visibleTooltip,
  changeValue,
  onSubmit,
  loading,
}) => (
  <Container>
    <Main>
      <Title>Painel de Controle</Title>

      <SwitchContainer>
        <Switch
          id="balance_transfer"
          checked={hasBalanceTransfer}
          name="balance_transfer"
          onChange={() =>
            onChangeTransfer("hasBalanceTransfer", !hasBalanceTransfer)
          }
        />
        {tooltipBalanceTransfer && (
          <Relative>
            <TooltipBalanceTransfer
              close={() =>
                onChangeTransfer("hasBalanceTransfer", !hasBalanceTransfer)
              }
              typeTooltip="hasBalanceTransfer"
            />
          </Relative>
        )}
        <TextWrapper>
          <Text>Habilitar a transferência de saldo.</Text>
          <ButtonBalanceTransfer onClick={() => showTooltip()}>
            Entenda a transferência de saldo
          </ButtonBalanceTransfer>
          {visibleTooltip && (
            <TooltipBalanceTransfer
              close={() => showTooltip()}
              typeTooltip="transferLimit"
            />
          )}
        </TextWrapper>
      </SwitchContainer>

      <Text>
        * Esta funcionalidade permite realizar a transferência apenas para o
        macefício de mesma titularidade.
      </Text>

      {hasBalanceTransfer && (
        <Fragment>
          <Text>Para concluir, escolha a modalidade:</Text>

          <SwitchContainer>
            <Switch
              id="foodToMeal"
              checked={hasFoodToMeal}
              name="foodToMeal"
              onChange={() => onChangeTransfer("hasFoodToMeal", !hasFoodToMeal)}
            />
            <Text>
              Autorizo a transferência de saldo do cartão{" "}
              <strong>alimentação para refeição.</strong>
            </Text>
          </SwitchContainer>

          <SwitchContainer>
            <Switch
              id="mealToFood"
              checked={hasMealToFood}
              name="mealToFood"
              onChange={() => onChangeTransfer("hasMealToFood", !hasMealToFood)}
            />
            <Text>
              Autorizo a transferência de saldo do cartão{" "}
              <strong>refeição para alimentação.</strong>
            </Text>
          </SwitchContainer>
        </Fragment>
      )}
    </Main>
    <Main>
      <Title>Valor mínimo de utilização</Title>
      <Text>Defina um valor mínimo de uso para macefício.</Text>

      <InputContainer>
        <InputWrapper>
          <span>R$</span>
          <CustomField
            label="Alimentação"
            value={foodValue}
            maxLength="12"
            disabled={!hasFoodToMeal}
            onChange={({ target }) => changeValue("foodValue", target.value)}
          />
        </InputWrapper>
        <InputWrapper>
          <span>R$</span>
          <CustomField
            label="Refeição"
            value={mealValue}
            maxLength="12"
            disabled={!hasMealToFood}
            onChange={({ target }) => changeValue("mealValue", target.value)}
          />
        </InputWrapper>
      </InputContainer>

      <Text>
        * O não preenchimento indica que os colaboradores podem transferir o
        valor integral de cada macefício.
      </Text>

      <CheckBox
        label="Concordo com as habilitações de transferência de saldo entre os 
        macefícios dos cartões mac Refeição e Alimentação, de todos os meus 
        colaboradores, conforme acordo sindical da minha categoria isentando a 
        mac de qualquer responsabilidade."
        checked={hasAcceptTerms}
        onChange={() => onChangeTransfer("hasAcceptTerms", !hasAcceptTerms)}
        name="accept_terms"
        id="accept_terms"
        disabled={false}
      />

      <ButtonsWrapper>
        <Button
          id="button_submit"
          onPress={() => onSubmit()}
          value="Salvar Alteração"
          buttonType={buttonTypesEnum.primary}
          type="submit"
          disabled={!hasAcceptTerms}
          loading={loading}
        />
      </ButtonsWrapper>
    </Main>
  </Container>
);

BalanceTransfer.propTypes = {
  hasBalanceTransfer: bool.isRequired,
  hasFoodToMeal: bool.isRequired,
  hasMealToFood: bool.isRequired,
  onChangeTransfer: func.isRequired,
  hasAcceptTerms: bool.isRequired,
  tooltipBalanceTransfer: bool.isRequired,
  mealValue: string.isRequired,
  foodValue: string.isRequired,
  showTooltip: func.isRequired,
  visibleTooltip: bool.isRequired,
  changeValue: func.isRequired,
  onSubmit: func.isRequired,
  loading: bool.isRequired,
};

export default BalanceTransfer;
