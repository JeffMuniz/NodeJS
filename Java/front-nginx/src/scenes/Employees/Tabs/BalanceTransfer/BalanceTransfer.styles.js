import styled from "styled-components";
import { veryLightBlack, blue } from "@colors";

const PBase = styled.p`
  font-style: normal;
  letter-spacing: 0.1px;
  font-weight: normal;
  margin-top: 1em;
  margin-bottom: 1em;
`;

export const Container = styled.section`
  display: flex;
  justify-content: space-between;
  width: 100%;
`;

export const Relative = styled.div`
  position: relative;
`;

export const Main = styled.div`
  flex: 1;
  display: flex;
  flex-direction: column;

  &:first-child {
    padding-right: 2em;
  }
`;

export const Title = styled.span`
  font-weight: bold;
  font-size: 18px;
  letter-spacing: -0.3px;
  color: ${veryLightBlack};
`;

export const SwitchContainer = styled.div`
  display: flex;
  align-items: center;
  margin-top: 1em;
`;

export const Text = styled(PBase)`
  line-height: 16px;
  font-size: 14px;
`;

export const ButtonsWrapper = styled.div`
  display: flex;
  justify-content: flex-end;
  width: 97%;
  margin-top: 1em;
`;

export const InputContainer = styled.div`
  display: flex;
  width: 100%;
  justify-content: space-between;
`;

export const InputWrapper = styled.div`
  display: flex;
  align-items: center;

  > span {
    margin-top: 10px;
    margin-right: 10px;
  }
`;

export const ButtonBalanceTransfer = styled.span`
  font-style: normal;
  font-weight: 600;
  font-size: 12px;
  line-height: 24px;
  margin-right: 5px;

  color: ${blue};
  cursor: pointer;
`;

export const TextWrapper = styled.div`
  display: flex;
  flex-direction: column;
  position: relative;

  > p {
    margin-bottom: 2px;
  }
`;
