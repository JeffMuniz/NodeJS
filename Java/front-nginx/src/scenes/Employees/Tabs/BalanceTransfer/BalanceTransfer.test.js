import React from "react";
import ShallowRenderer from "react-test-renderer/shallow";
import BalanceTransfer from "./BalanceTransfer";

jest.mock("@enums", () => ({
  buttonTypes: { primary: "" },
}));

jest.mock("@common", () => ({
  CheckBox: "CheckBox",
  Switch: "Switch",
  Button: "Button",
}));

const defaultProps = {
  hasBalanceTransfer: false,
  hasFoodToMeal: false,
  hasMealToFood: false,
  hasAcceptTerms: false,
  onChangeTransfer: jest.fn(),
  mealValue: "0,00",
  foodValue: "0,00",
  showPopover: jest.fn(),
  visiblePopover: false,
};

describe("Balance Transfer - Scene", () => {
  it("Should render it", () => {
    // given
    const renderer = new ShallowRenderer();

    // when
    renderer.render(<BalanceTransfer {...defaultProps} />);
    const result = renderer.getRenderOutput();

    // then
    expect(result).toMatchSnapshot();
  });
});
