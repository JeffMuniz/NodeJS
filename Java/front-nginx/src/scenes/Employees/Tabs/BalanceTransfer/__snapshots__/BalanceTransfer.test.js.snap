// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Balance Transfer - Scene Should render it 1`] = `
<styled.section>
  <styled.div>
    <styled.span>
      Painel de Controle
    </styled.span>
    <styled.div>
      <Switch
        checked={false}
        id="balance_transfer"
        name="balance_transfer"
        onChange={[Function]}
      />
      <styled.div>
        <Styled(styled.p)>
          Habilitar a transferência de saldo.
        </Styled(styled.p)>
        <styled.span
          onClick={[Function]}
        >
          Entenda a transferência de saldo
        </styled.span>
      </styled.div>
    </styled.div>
    <Styled(styled.p)>
      * Esta funcionalidade permite realizar a transferência apenas para o macefício de mesma titularidade.
    </Styled(styled.p)>
  </styled.div>
  <styled.div>
    <styled.span>
      Valor mínimo de utilização
    </styled.span>
    <Styled(styled.p)>
      Defina um valor mínimo de uso para macefício.
    </Styled(styled.p)>
    <styled.div>
      <styled.div>
        <span>
          R$
        </span>
        <CustomField
          component="input"
          disabled={true}
          errors={Object {}}
          iconErro={false}
          id={null}
          label="Alimentação"
          mask={null}
          maxLength="12"
          onChange={[Function]}
          placeholder={null}
          touched={Object {}}
          value="0,00"
          width={null}
        />
      </styled.div>
      <styled.div>
        <span>
          R$
        </span>
        <CustomField
          component="input"
          disabled={true}
          errors={Object {}}
          iconErro={false}
          id={null}
          label="Refeição"
          mask={null}
          maxLength="12"
          onChange={[Function]}
          placeholder={null}
          touched={Object {}}
          value="0,00"
          width={null}
        />
      </styled.div>
    </styled.div>
    <Styled(styled.p)>
      * O não preenchimento indica que os colaboradores podem transferir o valor integral de cada macefício.
    </Styled(styled.p)>
    <CheckBox
      checked={false}
      disabled={false}
      id="accept_terms"
      label="Concordo com as habilitações de transferência de saldo entre os  macefícios dos cartões mac Refeição e Alimentação, de todos os meus  colaboradores, conforme acordo sindical da minha categoria isentando a  mac de qualquer responsabilidade."
      name="accept_terms"
      onChange={[Function]}
    />
    <styled.div>
      <Button
        buttonType=""
        disabled={true}
        id="button_submit"
        loading={undefined}
        onPress={[Function]}
        type="submit"
        value="Salvar Alteração"
      />
    </styled.div>
  </styled.div>
</styled.section>
`;
